packages/userstorage/src/browser/user-storage-service-filesystem.ts

```diff
@@ -0,0 +1,110 @@
+                this.toDispose.push(disposable)
+            );
+            this.toDispose.push(this.watcher.onFilesChanged(changes => this.onDidFilesChanged(changes)));
+            return new URI(home.uri).resolve(THEIA_USER_STORAGE_FOLDER);
+        });
+    }
+
+    dispose(): void {
+        this.toDispose.dispose();
+    }
+
+    async onDidFilesChanged(fileChanges: FileChange[]): Promise<void> {
+        const uris: string[] = [];
+        for (const change of fileChanges) {
+            const userStorageUri = await this.toUserStorageUri(change.uri);
+            uris.push(userStorageUri.toString());
+        }
+        this.onUserStorageChangedEmitter.fire({ uris });
+    }
+
+    readContents(uri: string) {
+        return this.userStorageFolder
+            .then(async folderUri => {
+                const filesystemUri = await this.toFilesystemURI(new URI(uri));
+                return this.fileSystem.resolveContent(filesystemUri.toString());
+            })
+            .then(({ stat, content }) => content);
+    }
+
+    saveContents(uri: string, content: string) {
+
+        return this.userStorageFolder.then(async folderUri => {
+            const filesystemUri = await this.toFilesystemURI(new URI(uri));
+
+            this.fileSystem.getFileStat(filesystemUri.toString()).then(fileStat => {
+                this.fileSystem.setContent(fileStat, content).then(() => Promise.resolve());
+            });
+        });
+    }
+
+    get onUserStorageChanged(): Event<UserStorageChangeEvent> {
+        return this.onUserStorageChangedEmitter.event;
+    }
+
+    /**
+     * Creates a new user storage URI from the filesystem URI.
+     * @param fsPath The filesystem URI
+     */
+    toUserStorageUri(rawUri: URI): Promise<URI> {
+        return this.userStorageFolder.then(folder => {
+            const userStorageRelativePath = this.getRelativeUserStoragePath(folder, rawUri);
+            return new URI('').withScheme(UserStorageUri.SCHEME).withPath(userStorageRelativePath).withFragment(rawUri.fragment).withQuery(rawUri.query);
+        });
+    }
+
+    /**
+     * Returns the path relative to the user storage filesystem uri i.e if the user storage root is
+     * 'file://home/user/.theia' and the fileUri is 'file://home/user.theia/keymaps.json' it will return 'keymaps.json'
+     * @param userStorageURI User storage Root URI
+     * @param fileUri User storage
+     */
+    getRelativeUserStoragePath(userStorageURI: URI, fileUri: URI): string {
+        /* + 1 so that it removes the beginning slash  i.e return keymaps.json and not /keymaps.json */
+        return fileUri.toString().slice(userStorageURI.toString().length + 1);
+    }
+
+    /**
+     * Returns the associated filesystem URI from the user storage argument passed as argument
+     * @param userStorageUri User storage URI to be converted in filesystem URI
+     */
+    toFilesystemURI(userStorageUri: URI): Promise<URI> {
```

Comment:
This method doesn't return a `file://`-URI but it should.
Also, if you were making these methods (toUserStorageUri and this one) static and pass the userStorageFolder as an additional argument, you wouldn't need such a test setup (and mocks), but could simply test them with different args.

Reply:
It does return a `file://` uri though

`userStorageUri.withPath(root.path.join(userStorageUri.path.toString()))`

because in this `root` is a file:// -URI
Ok that makes sense for the static methods, although I feel like keeping all the mocks as I still want to test the implementation as a whole (testing registering a client and whatnot).

Reply:
but the scheme is still the user storage one (see your test assertions)

Reply:
Ah you're right, my bad :)